cli/src/graphar_cli/config.py

```diff
@@ -0,0 +1,253 @@
+        if self.is_primary and self.nullable:
+            msg = f"Primary key `{self.name}` must not be nullable."
+            raise ValueError(msg)
+        if self.is_primary:
+            self.nullable = False
+        elif self.nullable is None:
+            self.nullable = True
+        return self
+
+
+class PropertyGroup(BaseModel):
+    properties: List[Property]
+    file_type: Optional[Literal["parquet", "orc", "csv", "json"]] = None
+
+    @field_validator("properties")
+    def check_properties_length(cls, v):
+        if len(v) == 0:
+            msg = "properties must not be empty."
+            raise ValueError(msg)
+        return v
+
+
+class Source(BaseModel):
+    file_type: Optional[Literal["parquet", "orc", "csv", "json"]] = None
+    path: str
+    delimiter: str = ","
+    columns: Dict[str, str]
+
+    @field_validator("path")
+    def check_path(cls, v):
+        path = Path(v).resolve().absolute()
+        if not path.is_file():
+            msg = f"'{path}' is not a file."
+            raise ValueError(msg)
+        return v
+
+    @field_validator("delimiter")
+    def check_delimiter(cls, v):
+        if len(v) != 1:
+            msg = "delimiter must be a single character."
+            raise ValueError(msg)
+        return v
+
+    @model_validator(mode="after")
+    def check_file_type(self) -> Self:
+        if not self.file_type:
+            file_type = Path(self.path).suffix.removeprefix(".")
+            if file_type == "":
+                msg = f"File {self.path} has no file type suffix"
+                raise ValueError(msg)
+            if file_type not in support_file_types:
+                msg = f"File type {file_type} not supported"
+                raise ValueError(msg)
+            self.file_type = file_type
+        return self
+
+
+class Vertex(BaseModel):
+    type: str
+    labels: List[str] = []
+    chunk_size: Optional[int] = None
+    validate_level: Optional[Literal["no", "weak", "strong"]] = None
+    prefix: Optional[str] = None
+    property_groups: List[PropertyGroup]
+    sources: List[Source]
+
+    @field_validator("property_groups")
+    def check_property_groups_length(cls, v):
+        if len(v) == 0:
+            msg = "property_groups must not be empty."
+            raise ValueError(msg)
+        return v
+
+    @field_validator("sources")
+    def check_sources_length(cls, v):
+        if len(v) == 0:
+            msg = "sources must not be empty."
+            raise ValueError(msg)
+        return v
+
+    @model_validator(mode="after")
+    def check_vertex_prefix(self) -> Self:
+        prefix = self.prefix
+        type = self.type
+        if not prefix:
+            self.prefix = f"vertex/{type}/"
+        return self
+
+
+class AdjList(BaseModel):
+    ordered: bool
+    aligned_by: Literal["src", "dst"]
+    file_type: Literal["parquet", "orc", "csv", "json"]
+
+
+class Edge(BaseModel):
+    edge_type: str
+    src_type: str
+    src_prop: Optional[str] = None
```

Comment:
what are the `src_prop`and `dst_prop` mean here?

Reply:
They mean the props in `src_vertex` and `dst_vertex`. We use them to find the `vertex_id`. They should not be `Optional`. I will change them.